### PR TITLE
Add calibration demo notebook

### PR DESCRIPTION
**Pull request by georgeyw**
Requesting primary review from @jqhoogland, but @edmundlth please also check the comments in the notebook for accuracy if you have time! I am more basing this off of what I've heard people say than from some internal theoretical foundation. 

(adding @svwingerden for awareness / optional review)

**Main question for this PR from me:**
Do we want the actual code to run and plot sweeps (which I think would mostly be a `for` loop plus `plot_sweep()` in the notebook) to get checked into the repo, or is it enough for it to live in a demo notebook like this? I think I'm leaning no on this one (compared to leaning yes on checking in `plot_trace()` for the diagnostics demo notebook). Most of the thought there is on little plotting details changing around (e.g. some cases you'd only be checking one model and don't need a legend at all, others you might have dozens and want the color spectrum legend for different $log(M)$ values as with the ICL calibration) that might be annoying to catch all the edge cases of (plus discount factor on lines of code due to our codebase not being very stable yet).

### CODE REVIEW

**Comment by georgeyw** [comment]
Oh I also hadn't seen the initial spike before in a calibration sweep. I think I'm inclined to believe it's not a problem if the graph converges eventually, but please correct me if I'm wrong on that.

**Review by edmundlth — COMMENTED** [review]
## Review notes (discussed in our call)
My understanding is that this notebook is a demo calibration process for **large** models where we don't have ground truth $\lambda$. It demonstrates that there are still quite a lot of diagnostic process that make us more confident that $\hat{\lambda}$ reflect ordering of degeneracy within a training run. Perhaps this point might need to be emphasised a little in prose in the notebook. 

Other items: 
- Axis labels in the `plot_sweep()` cell and write the interpretation of those plots + best practice (can change in the future) recommendation for $\epsilon, \gamma$ based on those plots. 
- On the assumption that this doesn't need to be run often, run the training for longer and plot the train/test loss to see that they converged. 
- Run the SGLD chains for a lot longer so that one can see the compute-vs-$\epsilon$ tradeoff in the $\hat{\lambda}$-vs-SGLD-steps graphs. That's valuable information for calibrating hyperparameter too. 
- OOM of Num SGD step count in the graphs doesn't look right. 
- Might want to add in the loss trace plots for different combination of $\epsilon, \gamma$ as calibration information as well. 


## Responding to the comments at the end of notebook

> The top two graphs don't appear to converge within the 200 draw steps.(200 also isn't a lot, you may want to run your sweeps with more than 200 draw steps)

Yes, 200 is not enough based on those graphs. For this demo, plots with long SGLD chain would give us some information about how many steps are needed to converge for small $\epsilon$. 

> The top right graph has a small negative dip, but that isn't too concerning.

The loss trace should catch this kind of instability. But model that hasn't been trained for long enough frequently have this SGLD-becomes-training-and-not-sampling issue. 

> The bottom left graph has lines that appear to cross, but the lines are pretty close to overlapping.

Not sure how to interpret this. Filing this away until it occur in other settings again. 

> The bottom two graphs do appear to converge, and either one might be fine for parameter selection, but it may also be worth running a new sweep with higher learning rate.

For this demo, yes, I would agree, just to document possible behaviour (e.g. might get exploding nonsense for large enough epsilon).

**Comment by edmundlth** [comment]
> Oh I also hadn't seen the initial spike before in a calibration sweep. I think I'm inclined to believe it's not a problem if the graph converges eventually, but please correct me if I'm wrong on that.

I was just about to mention this when I see this comment for the first time. Sorry I missed it earlier. 

I think that is **NOT** a healthy sign. Base on this curve shape, I would say that the loss trace should look like an initial increase ("thermalisation") and then drop below the initial loss in a sudden discovery of better solution (i.e. SGLD becomes training again). If that is true, this related to 
- loss trace should definitely be part of the calibration / diagnosis tool box. 
- the network hasn't been trained to some stable equilibrium
- non-negative lambdahat doesn't mean healthy, we need samples to not consistently drop far below initial loss to believe that low lambdahat $\implies$ low local complexity

**Comment by georgeyw** [comment]
Thanks for all the feedback Edmund, I think this should be updated now to reflect all the comments. Let me know if it looks more presentable now and if there's anything else that I might have missed!

**Review by jqhoogland — CHANGES_REQUESTED** [review]
Approved after we get Edmund to weigh in about thermalization. Maybe also another comment about the spikes (see discussion in Discord #theory channel)

**Review by edmundlth — APPROVED** [review]
Thanks for the update @georgeyw . It looks good. Just a few notes: 
* I think you meant `1e-5` at the end when you're suggesting hyperparam based on loss traces? The `1e-3`'s all blows up. 
* Not sure I want to call the spikes, initial or otherwise, "thermalisation". I don't really know what they are (probably something about highly curved parts of the loss landscape?). In my experiments, they occur at some high epsilons. I think they are not healthy signs, but probably not cause to call the SGLD run fail. In short, I don't know. 
* I also don't know if monotonically increasing lambdahat-vs-numsteps curve is the most desirable. I can see that no being true for some toy small systems simply due to noise. But perhaps in larger systems .... Not sure. Sorry about the whole lot of not sures. This is why I am picturing a "dashboard" system: many things could go wrong that we don't know about, and each might have a different signal. E.g. the sharp rise and drop of your earlier notebook is a cause for concern and investigation if this is a true production run. E.g. perhaps another panel of the dashboard could clarify the picture, I think distance from initialisation over SGLD steps could be valuable for answering this particular question.

**Comment by georgeyw** [comment]
Yeah good catch, it should be `1e-5`

I think maybe it would be good then to include more of this uncertainty in the prose of the notebook? It seems like the message is that this is more an art than a science for now. I think it is also probably ok to not expect a reader of this notebook to walk away with all their questions answered.

**Comment by edmundlth** [comment]
Yes. Perhaps even at the very beginning. 
* Many uncertainty and unknown unknowns. 
* More art than science is exactly right for now, in the same way SGD/ADAM training is more art than science at the moment, just a very well developed art from many generations of natural selection. 
* Getting our own selection pressure by collecting more signals of health is what we need at the moment.

**Review by jqhoogland — APPROVED** [review]
